File: TheWatch-Android/app/src/main/java/com/thewatch/app/service/QuickTapCoordinator.kt
// QuickTapCoordinator — app-level coordinator for tap-to-SOS.
// Collects QuickTapEvents from QuickTapDetector and dispatches to SOSService.
// Runs at application scope, independent of any screen.

package com.thewatch.app.service

import android.content.Context
import android.content.Intent
import android.util.Log
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Application-scoped coordinator for quick-tap SOS triggering.
 *
 * Flow:
 * 1. User rapidly taps volume button (4x in 5s)
 * 2. QuickTapDetector emits QuickTapEvent
 * 3. This coordinator catches it, escalates location, fires SOSService
 * 4. LocationCoordinator switches to EMERGENCY mode
 */
@Singleton
class QuickTapCoordinator @Inject constructor(
    @ApplicationContext private val context: Context,
    private val quickTapDetector: QuickTapDetector,
    private val locationCoordinator: LocationCoordinator
) {
    companion object {
        private const val TAG = "QuickTapCoordinator"
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private val _isSOSActive = MutableStateFlow(false)
    val isSOSActive: StateFlow<Boolean> = _isSOSActive.asStateFlow()

    private val _lastTapEvent = MutableStateFlow<QuickTapEvent?>(null)
    val lastTapEvent: StateFlow<QuickTapEvent?> = _lastTapEvent.asStateFlow()

    init {
        // Collect tap events at app scope
        scope.launch {
            quickTapDetector.tapEvents.collect { event ->
                handleTapEvent(event)
            }
        }
    }

    fun enable() {
        quickTapDetector.enable()
        Log.i(TAG, "Quick-tap SOS enabled")
    }

    fun disable() {
        quickTapDetector.disable()
        Log.i(TAG, "Quick-tap SOS disabled")
    }

    private fun handleTapEvent(event: QuickTapEvent) {
        _lastTapEvent.value = event
        _isSOSActive.value = true

        Log.i(TAG, "SOS triggered via ${event.triggerType}: ${event.tapCount} taps in ${event.windowMs}ms")

        // Escalate location to emergency mode
        locationCoordinator.escalateToEmergency()

        // Fire SOS service
        val intent = Intent(context, SOSService::class.java).apply {
            action = SOSService.ACTION_START_SOS
            putExtra("alert_type", "Emergency")
            putExtra("description", "Quick-tap SOS: ${event.tapCount} ${event.triggerType.name} taps")
            putExtra("trigger_source", "QUICK_TAP")
        }
        context.startForegroundService(intent)
    }

    /**
     * Cancel SOS triggered by quick-tap.
     * Called from UI or by clear-word phrase detection.
     */
    fun cancelSOS() {
        if (_isSOSActive.value) {
            _isSOSActive.value = false
            locationCoordinator.deescalateToNormal()
            quickTapDetector.reset()

            val intent = Intent(context, SOSService::class.java).apply {
                action = SOSService.ACTION_CANCEL_SOS
            }
            context.startService(intent)
            Log.i(TAG, "Quick-tap SOS cancelled")
        }
    }
}

File: TheWatch-iOS/Services/QuickTapCoordinator.swift
// QuickTapCoordinator — app-level coordinator for tap-to-SOS.
// Collects QuickTapEvents from QuickTapDetector and dispatches to SOS pipeline.
// Runs at application scope, independent of any screen.

import Foundation
import Combine

/// Application-scoped coordinator for quick-tap SOS triggering.
///
/// Flow:
/// 1. User rapidly taps volume button (4x in 5s) or shakes device
/// 2. QuickTapDetector emits QuickTapEvent
/// 3. This coordinator catches it, escalates location, fires SOS
/// 4. LocationCoordinator switches to EMERGENCY mode
@Observable
final class QuickTapCoordinator {
    static let shared = QuickTapCoordinator()

    private(set) var isSOSActive = false
    private(set) var lastTapEvent: QuickTapEvent?

    private let quickTapDetector = QuickTapDetector.shared
    private let locationCoordinator = LocationCoordinator.shared
    private let sosTriggerService = SosTriggerService.shared
    private var cancellables = Set<AnyCancellable>()

    /// Whether quick-tap SOS is enabled by the user.
    var isEnabled: Bool {
        get { UserDefaults.standard.bool(forKey: "quickTapSOSEnabled") }
        set {
            UserDefaults.standard.set(newValue, forKey: "quickTapSOSEnabled")
            if newValue { enable() } else { disable() }
        }
    }

    private init() {
        // Collect tap events at app scope
        quickTapDetector.tapEventPublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] event in
                self?.handleTapEvent(event)
            }
            .store(in: &cancellables)

        // Auto-enable if user previously enabled
        if UserDefaults.standard.bool(forKey: "quickTapSOSEnabled") {
            quickTapDetector.enable()
        }
    }

    func enable() {
        quickTapDetector.enable()
        print("[QuickTapCoordinator] Quick-tap SOS enabled")
    }

    func disable() {
        quickTapDetector.disable()
        print("[QuickTapCoordinator] Quick-tap SOS disabled")
    }

    private func handleTapEvent(_ event: QuickTapEvent) {
        lastTapEvent = event
        isSOSActive = true

        print("[QuickTapCoordinator] SOS triggered via \(event.triggerType.rawValue): " +
              "\(event.tapCount) taps in \(event.windowSeconds)s")

        // Trigger SOS through the centralized service — handles auth, API, offline queue
        sosTriggerService.trigger(
            source: .quickTap,
            description: "Quick-tap SOS: \(event.tapCount) \(event.triggerType.rawValue) taps"
        )

        // Fire SOS notification — any active ViewModel can observe
        NotificationCenter.default.post(
            name: .quickTapSOS,
            object: nil,
            userInfo: [
                "type": "emergency",
                "description": "Quick-tap SOS: \(event.tapCount) \(event.triggerType.rawValue) taps",
                "trigger_source": "QUICK_TAP",
                "tap_count": event.tapCount
            ]
        )
    }

    /// Cancel SOS triggered by quick-tap.
    func cancelSOS() {
        guard isSOSActive else { return }
        isSOSActive = false
        locationCoordinator.deescalateToNormal()
        quickTapDetector.reset()

        NotificationCenter.default.post(
            name: .quickTapSOSCancelled,
            object: nil
        )
        print("[QuickTapCoordinator] Quick-tap SOS cancelled")
    }
}

// MARK: - Notification Names

extension Notification.Name {
    static let quickTapSOS = Notification.Name("com.thewatch.quickTapSOS")
    static let quickTapSOSCancelled = Notification.Name("com.thewatch.quickTapSOSCancelled")
}
